lib/rubocop/cop/lint/missing_super.rb

```diff
@@ -0,0 +1,99 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Lint
+      # This cop checks for the presence of constructors and lifecycle callbacks
+      # without calls to `super`.
+      #
+      # @example
+      #   # bad
+      #   class Employee < Person
+      #     def initialize(name, salary)
+      #       @salary = salary
+      #     end
+      #   end
+      #
+      #   # good
+      #   class Employee < Person
+      #     def initialize(name, salary)
+      #       super(name)
+      #       @salary = salary
+      #     end
+      #   end
+      #
+      #   # bad
+      #   class Parent
+      #     def self.inherited(base)
+      #       do_something
+      #     end
+      #   end
+      #
+      #   # good
+      #   class Parent
+      #     def self.inherited(base)
+      #       super
+      #       do_something
+      #     end
+      #   end
+      #
+      class MissingSuper < Base
+        CONSTRUCTOR_MSG = 'Call `super` to initialize state of the parent class.'
+        CALLBACK_MSG    = 'Call `super` to invoke callback defined in the parent class.'
+
+        STATELESS_CLASSES = %w[BasicObject Object].freeze
+
+        OBJECT_LIFECYCLE_CALLBACKS  = %i[method_missing respond_to_missing?].freeze
+        CLASS_LIFECYCLE_CALLBACKS   = %i[inherited].freeze
+        MODULE_LIFECYCLE_CALLBACKS  = %i[included extended prepended].freeze
+
+        def on_def(node)
+          return unless offender?(node)
+
+          if node.method?(:initialize)
+            add_offense(node, message: CONSTRUCTOR_MSG) if inside_class_with_stateful_parent?(node)
+          elsif callback_method_def?(node)
+            add_offense(node, message: CALLBACK_MSG)
+          end
+        end
+
+        def on_defs(node)
+          return if !callback_method_def?(node) || contains_super?(node)
+
+          add_offense(node, message: CALLBACK_MSG)
+        end
+
+        private
+
+        def offender?(node)
+          (node.method?(:initialize) || callback_method_def?(node)) && !contains_super?(node)
+        end
+
+        def callback_method_def?(node)
+          method_name = node.method_name
+
+          if OBJECT_LIFECYCLE_CALLBACKS.include?(method_name) ||
+             CLASS_LIFECYCLE_CALLBACKS.include?(method_name)
+
+            node.each_ancestor(:class, :sclass, :module).first
+          elsif MODULE_LIFECYCLE_CALLBACKS.include?(method_name)
+            node.each_ancestor(:module).first
+          end
+        end
+
+        def contains_super?(node)
+          node.each_descendant(:super, :zsuper).any?
```

Comment:
`any` is not the same thing as `!.empty?`. It will work the same way in most cases, however this will fail if the array contains `nil`.

```ruby
[1] pry(main)> [nil].any?
=> false
[2] pry(main)> ![nil].empty?
=> true
```

Reply:
Correct, but `each_descendant(:super, :zsuper)` won't return array with nils - either empty or with `Node` instances.